main/modes/games/2048/mode_2048.c

```diff
@@ -0,0 +1,1296 @@
+ */
+static uint8_t getColor(uint32_t val);
+
+/**
+ * @brief Draws the grid, score, and tiles
+ *
+ */
+static void t48Draw(void);
+
+/**
+ * @brief Shows the title upon booting into the mode
+ *
+ * @param color Color of the title text
+ */
+static void t48StartScreen(uint8_t color);
+
+/**
+ * @brief Draws the final score and prompts the player to play again
+ *
+ * @param score The final value of the game
+ */
+static void t48DrawGameOverScreen(int64_t score);
+
+/**
+ * @brief Draw the win screen. It doesn't do anything else.
+ *
+ */
+static void t48DrawWinScreen(void);
+
+/**
+ * @brief Draw the screen asking to confirm they want to quit
+ *
+ */
+static void t48DrawConfirm(void);
+
+// LEDs
+
+/**
+ * @brief Automatically dims LEDs every frame
+ *
+ */
+static void t48DimLEDs(void);
+
+/**
+ * @brief Sets an LED to a color
+ *
+ * @param idx   Index of the LED
+ * @param color Color to set LED
+ */
+static void t48SetRGB(uint8_t idx, Color_t color);
+
+/**
+ * @brief Illuminate appropriate LEDs based on an indicated direction
+ *
+ * @param dir   Direction to illuminate LEDs
+ * @param color Color to set the LEDs to
+ */
+static void t48LightLEDs(Direction_t dir, Color_t color);
+
+/**
+ * @brief Based on the highest block value, set the LED color
+ *
+ * @return Color_t Color object send to the LEDs
+ */
+static Color_t t48GetLEDColors(void);
+
+/**
+ * @brief Get a random bright color for the LEDs
+ *
+ * @return Color_t Color to send to LEDs
+ */
+static Color_t t48RandColor(void);
+
+/**
+ * @brief Run an random LED program
+ *
+ */
+static void t48RandLEDs(void);
+
+/**
+ * @brief Returns values in hue sequence
+ *
+ * @return paletteColor_t Color to use
+ */
+static paletteColor_t t48Rainbow(void);
+
+// Audio
+/**
+ * @brief Restart audio when song ends
+ *
+ */
+static void t48BgmCb(void);
+
+//==============================================================================
+// Variables
+//==============================================================================
+
+const char modeName[]   = "2048";
+const char pressKey[]   = "Press any key to play";
+const char pressAB[]    = "Press A or B to reset the game";
+const char youWin[]     = "You got 2048!";
+const char continueAB[] = "Press A or B to continue";
+const char highScore[]  = "You got a high score!";
+const char paused[]     = "Paused!";
+const char pausedA[]    = "Press A to continue playing";
+const char pausedB[]    = "Press B to abandon game";
+
+const char highScoreKey[HS_COUNT][HS_KEYLEN]
+    = {"t48HighScore0", "t48HighScore1", "t48HighScore2", "t48HighScore3", "t48HighScore4"};
+const char highScoreInitialsKey[HS_COUNT][HS_KEYLEN] = {
+    "t48HSInitial0", "t48HSInitial1", "t48HSInitial2", "t48HSInitial3", "t48HSInitial4",
+};
+
+swadgeMode_t t48Mode = {
+    .modeName                 = modeName,
+    .wifiMode                 = NO_WIFI,
+    .overrideUsb              = false,
+    .usesAccelerometer        = false,
```

Comment:
One day

Reply:
I'm going to install my favorite flavor of linux on my desktop when I get a spare moment. I know how to get that one to work, and I know how to get the hardware to cooperate.

I just have to actually have the time to do it. Once that happens, tilt controls.

Reply:
I never got real Linux on my silly Dell, so I'm a wsl2 user. I have managed USB passthrough using usbipd though...